### PR TITLE
FLOPs calculation for Monty

### PR DESCRIPTION
**Pull request by hlee9212**

### CODE REVIEW

**Review by scottcanoe — APPROVED** [review]
LGTM.

**Comment by nielsleadholm** [comment]
Hey Hojae, just an initial comment before I go through in detail - can you please separate out the 2025 copyright changes as a separate PR? That will be a lot clearer than trying to combine it with the actual FLOPs-related changes.

**Comment by nielsleadholm** [comment]
Ok interesting, yeah I wonder whether it is worth making sure your branch is up-to-date with the `main` branch? Looking through the commit history, in early December there was

`Merge branch 'thousandbrainsproject:main' into hlee9212/flops`

But since then (4 days ago), the only merge call was:
`Merge branch 'hlee9212/flops' of github.com:hlee9212/monty_lab into hlee9212/flops`

i.e. the main branch hasn't been merged into your local branch since [Tristan's copyright PR ](https://github.com/thousandbrainsproject/monty_lab/pull/12) was merged. My recommendation would be trying that.

In general best practice is to do this frequently (i.e. fetch and merge the latest main into your local branches) - you're much less likely to run into a nasty merge conflict after working on a complex project for weeks. 

Let me know if you want help going through the steps for this.

File: monty_profiling/floppy/flop_counting/operations/arithmetic.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        try:
+            # Handle case where one input is scalar
+            if not all(shapes):
+                return shapes[1] if not shapes[0] else shapes[0]
+
+            # Create dummy arrays of ones with the given shapes
+            arrays = [np.ones(shape) for shape in shapes]
+            result = np.broadcast_arrays(*arrays)
+            return result[0].shape
+        except ValueError as e:
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/arithmetic.py:26]: Should we raise an error rather than a warning? Otherwise seems like it would be easy to overlook FLOPs that aren't being accounted for correctly.

File: monty_profiling/floppy/flop_counting/operations/arithmetic.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            warnings.warn(f"Invalid broadcast shape: {str(e)}")
+            return None
+
+    def count_flops(self, *args: Any, result: Any) -> Optional[int]:
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/arithmetic.py:30]: It would be helpful with more detailed doc strings for key functions like this one, e.g. something like the below:

```python
        """Count the FLOPs (Floating Point Operations) for the operation.

        This function estimates the number of FLOPs required to perform a numpy-like operation. It handles both scalar and array inputs
        and accounts for broadcasting by using the shape of the actual result.

        Args:
            *args (Any): Input arguments for the operation. Can include scalars or arrays.
            result (Any): The result of the operation, used to determine the output shape.

        Returns:
            Optional[int]: The estimated number of FLOPs, or None if the input is invalid.
        """
```

File: monty_profiling/floppy/flop_counting/operations/arithmetic.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            array_arg = args[1] if np.isscalar(args[0]) else args[0]
+            return np.size(array_arg)
+
+        # Handle array operations
+        arrays = [np.asarray(arg) for arg in args[:2]]
+        # Get the result shape directly from the actual result
+        # This ensures we correctly account for broadcasting
+        result_shape = np.asarray(result).shape
+
+        return np.prod(result_shape)
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/arithmetic.py:43]: Is it fair to say that line 41 and 43 is equivalent to the below? If nothing else, this could be a comment to help readers see it from a different vantage.

```python
return np.size(np.asarray(result))
```

File: monty_profiling/floppy/flop_counting/operations/linalg.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        Total: ~14n³ + 1 FLOPs
+        """
+        n = args[0].shape[0]
+        return 14 * n**3 + 1
+
+
+class InvOperation:
+    """FLOP count for np.linalg.inv operation."""
+
+
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/linalg.py:40]: Maybe delete some of this white space.

File: monty_profiling/floppy/flop_counting/operations/matmul.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+            # Compute broadcasted batch shape
+            batch_shape = self._compute_broadcast_batch_shape(a.shape, b.shape)
+            batch_size = np.prod(batch_shape) if batch_shape else 1
+
+            # Each element requires N multiplications and N-1 additions
+            # For each M×P elements in the result
+            return batch_size * M * P * (2 * N - 1)
+
+        except Exception as e:
+            warnings.warn(f"Error counting matmul FLOPs: {str(e)}")
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/matmul.py:82]: Again, might be better to raise an error rather than a warning.

File: monty_profiling/floppy/flop_counting/operations/matmul.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                # Vector dot product: 2N-1 FLOPs (N multiplications, N-1 additions)
+                return 2 * a.shape[0] - 1
+
+            if a.ndim == 1:  # 1D × 2D
+                M, N = 1, a.shape[0]
+                P = b.shape[1]
+            elif b.ndim == 1:  # 2D × 1D
+                M, N = a.shape
+                P = 1
+            else:  # ND × ND
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/matmul.py:68]: Maybe just add a comment that you index from the back due to potential batch dimensions.

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    "FloorDivideOperation",
+    "ModuloOperation",
+    "BitwiseAndOperation",
+    "BitwiseOrOperation",
+]
+class SumOperation:
+    """FLOP count for sum operation."""
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for sum operation."""
+        return np.size(args[0]) - 1
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:30]: I think this will depend if we're just summing a vector/summing along the first dimension, vs. summing along multiple dimensions?

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+        return np.size(args[0])
+
+
+class MinOperation:
+    """FLOP count for min operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for min operation.
+
+        Each element requires 1 comparison operation.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:61]: Maybe just for correctness, something like "Except for the first element?"

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+        return np.size(args[0]) - 1
+
+
+class MaxOperation:
+    """FLOP count for max operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for max operation.
+
+        Each element requires 1 comparison operation.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:72]: As above with min.

**Review by nielsleadholm — COMMENTED** [review]
Amazing amount of work Hojae, very impressive!

This is a "part 1" review, I haven't gone through everything, just to start getting some clarification on certain points while I finish going through it.

File: monty_profiling/floppy/flop_counting/operations/linalg.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+class CondOperation:
+    """FLOP count for np.linalg.cond operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for condition number calculation.
+
+        Condition number requires:
+        - SVD decomposition (~14n³ for an nxn matrix)
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/linalg.py:29]: Did you look up particular references for these elements? If so, it would be helpful to just link to them as a sanity check. Similar for e.g. `LU decomposition (~2/3 n³)`

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+        return np.size(args[0])  # (n-1) additions + 1 division
+
+
+class StdOperation:
+    """FLOP count for standard deviation operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for standard deviation operation.
+
+        Standard deviation requires:
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:155]: Really nicely documented!

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+            return 2 * n + 1  # 2n operations for weighted sum and 1 division
+        else:
+            # Unweighted average: sum + division
+            return n + 1  # n operations for sum and 1 division
+
+
+class LogOperation:
+    """FLOP count for logarithm operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:230]: I think this might be one that uses a Taylor series / look-up table, so we might want to have more than one FLOPs per element. 

https://math.stackexchange.com/questions/61209/what-algorithm-is-used-by-computers-to-calculate-logarithms

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+        """
+        return np.size(args[0])
+
+class PowerOperation:
+    """FLOP count for power operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for power operation.
+
+        Each element requires 1 power operation.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:243]: Wouldn't this be a function of the power? I think it would be `(k-1)n` FLOPs, where k is the power. E.g. if k=3, we do a*a*a = 2 FLOPs per element. I don't think any sum would be separate.

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+
+class FloorDivideOperation:
+    """FLOP count for floor divide operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for floor divide operation.
+
+        Each element requires 1 division operation.
+        """
+        return np.size(args[0])
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:256]: Would this be 2n? It feels like we are doing a division and then a round/min type call, i.e. two instances of n FLOPs.

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+
+
+class ModuloOperation:
+    """FLOP count for modulo operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for modulo operation.
+
+        Each element requires 1 modulo operation.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:265]: This also feels like a division followed by a comparitor, so 2n would be safer?

File: monty_profiling/floppy/flop_counting/operations/trigonometry.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+]
+
+class SineOperation:
+    """FLOP counter for sine operations."""
+
+    def count_flops(self, *args: Any, result: Any) -> Optional[int]:
+        """Count FLOPs for sine operation.
+
+        Sine is typically implemented using Taylor series or CORDIC algorithm.
+        We estimate 8 FLOPs per value based on common implementations.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/trigonometry.py:21]: The below image is the most common implementation I find. If we just take the two terms, to my count that is the below (10 FLOPs). I'm assuming that for low-number factorials, there is a look-up table rather than this explicitly being calculated.

1 FLOP : first subtraction of terms
2 FLOPs: x * x * x
1 FLOP: division by 3! (which is acquired by look-up)
1 FLOP: first addition of terms
4 FLOPs: x * x * x * x * x
1 FLOP: division by 5!

Total: 10 FLOPs

As discussed in the meeting, I think it would be safest to just set this to 10. I think it would also be worth including the above reasoning in the description for transparency, as we had for the others. 

Screenshot 2025-01-14 at 5 22 26 PM

File: monty_profiling/floppy/flop_counting/operations/trigonometry.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        """
+        if not self.validate_inputs(*args):
+            return None
+
+        return 8 * np.size(result)
+
+    def validate_inputs(self, *args: Any) -> bool:
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/trigonometry.py:28]: Can you clarify the specific need for this validate_inputs?

File: monty_profiling/floppy/flop_counting/operations/trigonometry.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+
+
+class CosineOperation:
+    """FLOP counter for cosine operations."""
+
+    def count_flops(self, *args: Any, result: Any) -> Optional[int]:
+        """Count FLOPs for cosine operation.
+
+        Cosine is typically implemented using Taylor series or CORDIC algorithm.
+        We estimate 8 FLOPs per value based on common implementations.
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/trigonometry.py:45]: This is typically of the below form:

Screenshot 2025-01-14 at 5 28 30 PM

File: monty_profiling/floppy/flop_counting/operations/trigonometry.py [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+
+        For 3D vectors, cross product requires:
+        - 6 multiplications
+        - 3 subtractions
+        Total: 9 FLOPs per cross product
+        """
+        # Get number of cross products being computed
+        num_operations = max(
+            1, result.shape[0] if isinstance(result, np.ndarray) else 1
+        )
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/trigonometry.py:76]: It's probably worth having a validation that we are operating with 3D vectors.

File: monty_profiling/floppy/flop_counting/operations/trigonometry.py [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+        return 9 * num_operations
+
+class ArccosOperation:
+    """FLOP counter for inverse cosine operations."""
+
+    def count_flops(self, *args: Any, result: Any) -> Optional[int]:
+        """Count FLOPs for inverse cosine operation.
+
+        Arccos is typically implemented using a combination of
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/trigonometry.py:85]: What was the source of this?

**Review by nielsleadholm — COMMENTED** [review]
Pt 2 comments, focusing on the `operations` FLOPs estimates.

File: monty_profiling/floppy/monty_flop_tracer.py [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+
+    def _log_flops(self, method_name, flops):
+        """Log FLOP counts to CSV file."""
+        with open(self.log_path, "a", newline="") as f:
+            writer = csv.writer(f)
+            writer.writerow([time.time(), self.current_episode, method_name, flops])
+
+    def _wrap_methods(self):
+        """Wrap key Monty methods to count FLOPs."""
+        self._original_methods = {
```
Comment (nielsleadholm) [inline monty_profiling/floppy/monty_flop_tracer.py:44]: Is there a reason this is defined here rather than in `init`?

File: monty_profiling/floppy/monty_flop_tracer.py [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+                    result = original(*args, **kwargs)
+
+                step_flops = self.flop_counter.flops
+                self._log_flops(method_name, step_flops)
+                self.total_flops += step_flops
+                return result
+
+            setattr(target_obj, method_name, wrapped)
+
+        # Special wrapper for run_episode to track total FLOPs independently
```
Comment (nielsleadholm) [inline monty_profiling/floppy/monty_flop_tracer.py:76]: I'm curious, was this equal/very close to the sum of the other, inner methods? `self.monty.step` contains both `_matching_step` and `_exploratory_step`, and to my eye, all three of these are tracked and will contribute to `self.total_flops`, which makes me concerned there may be some over counting?

File: monty_profiling/frameworks/models/evidence_matching.py [hunk 23]
```diff
@@ -0,0 +1,316 @@
...
+        # Execute KDTree query
+        vote_location_tree = KDTree(
+            graph_location_vote,
+            leafsize=40,
+        )
+        vote_nn = 3  # TODO: Make this a parameter?
+        if graph_location_vote.shape[0] < vote_nn:
+            vote_nn = graph_location_vote.shape[0]
+
+        # Account for FLOPs for KDTree Construction
```
Comment (nielsleadholm) [inline monty_profiling/frameworks/models/evidence_matching.py:107]: Do you have a reference for what this was based on that we can link to?

File: monty_profiling/frameworks/models/evidence_matching.py [hunk 24]
```diff
@@ -0,0 +1,316 @@
...
+            k=vote_nn,
+            p=2,
+            workers=1,
+        )
+
+        # Account for FLOPs for KDTree query
+        num_search_points = len(self.possible_locations[graph_id])
+        num_reference_points = len(graph_location_vote)
+
+        # Tree Traversal FLOPs
```
Comment (nielsleadholm) [inline monty_profiling/frameworks/models/evidence_matching.py:123]: Similarly for this?

File: monty_profiling/floppy/flop_counting/operations/linalg.py (hunk 10 above)
Comment (nielsleadholm) [reply monty_profiling/floppy/flop_counting/operations/linalg.py:29]: Ok perfect, thanks for the detailed response! Is it possible to then add to the docstring something like

```
An estimate for complexity of SVD is 
~2mn^2  + 11n^3
per equation 11.22 in "Numerical Linear Algebra" by Trefethen and Bau

Here we assume m=n (a square matrix), and so estimate FLOPs as
 14n^3
```

And good to know about efficiency of different operations, but agree it's not a priority at the moment given how often we use SVD.

File: monty_profiling/floppy/flop_counting/operations/linalg.py [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+
+
+class InvOperation:
+    """FLOP count for np.linalg.inv operation."""
+
+    def count_flops(self, *args: Any, result: Any) -> int:
+        """Count FLOPs for matrix inversion.
+
+        Matrix inversion using LU decomposition:
+        - LU decomposition (~2/3 n³)
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/linalg.py:44]: It would be great if, like for SVD, we could have a simple reference here for `InvOperation` and `EigOperation`

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py (hunk 7 above)
Comment (nielsleadholm) [reply monty_profiling/floppy/flop_counting/operations/numpy_functions.py:30]: Yeah good point about upper-bounding, maybe just add a comment to the doc-string to make that clear, i.e. we ignore reduce from summing along specific axes.

File: monty_profiling/floppy/flop_counting/operations/arithmetic.py [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class Multiplication(ArithmeticOperation):
+    """Class for multiplication operation."""
+
+    def __init__(self):
+        super().__init__("multiply")
+
+
+class Division(ArithmeticOperation):
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/arithmetic.py:67]: I noticed while doing some further reading on FLOPs that division is sometimes considered >1 (e.g. 4) FLOPs, due to differences in how it is handled in hardware. However, it is also often estimated as 1. I think it would be worth adding a reference somewhere to the following as justification for our use of 1. 

https://www.stat.cmu.edu/~ryantibs/convexopt-F18/scribes/Lecture_19.pdf

File: monty_profiling/floppy/flop_counting/operations/numpy_functions.py [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+
+        Each element requires 1 bitwise and operation.
+        """
+        return np.size(args[0])
+
+
+class BitwiseOrOperation:
+    """FLOP count for bitwise or operation."""
+
+
```
Comment (nielsleadholm) [inline monty_profiling/floppy/flop_counting/operations/numpy_functions.py:289]: Lots of white-space here.

File: monty_profiling/.gitignore [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+*.py,cover
+.hypothesis/
+.pytest_cache/
+cover/
+
+# Translations
+*.mo
+*.pot
+
+# Django stuff:
```
[the remark this answers is not in the document]
Comment (nielsleadholm) [reply monty_profiling/.gitignore:65]: Ah ok fair enough!

File: monty_profiling/floppy/flop_counting/operations/linalg.py (hunk 10 above)
Comment (nielsleadholm) [reply monty_profiling/floppy/flop_counting/operations/linalg.py:29]: Re. the ord=2, it seems to depend on whether the input is a vector or matrix, see the table here (https://numpy.org/doc/2.1/reference/generated/numpy.linalg.norm.html)

Summarized:


Frobenius norm is performed when `ord=fro`

**Review by nielsleadholm — COMMENTED** [review]
Additional comments

**Review by nielsleadholm — COMMENTED** [review]
Pt 5/5 (done with first pass through everything). 

I loved all the unit tests at the end, just beautiful. 

A few high-level comments
- The documentation in the Readme is really good, although I think it would be useful to have some more details at the end describing the big picture, i.e. how `TrackedArray`, `FlopCounter`, `MontyFlopTracer` and e.g. `wrap_monty_with_flops` work together to track the FLOPs. You gave some great examples of how the first two work in isolation, but taking a step back to look at how we pass around the FLOP counters etc. to get an overall result regardless of the underlying functions would be helpful.
- I have not implemented context managers myself (which I imagine applies to some others), so it would be useful to have a brief description of how this works in Python, e.g. the importance of `__enter__`.
- It would be great if the PR description (i.e. for posterity/future reference) had some details on end-to-end checks you ran, in particular, i) that the sum of the separate Monty-step counters equals the high-level `run_episode` counter, and ii) that the accuracy and num-steps to convergence on a Monty experiment is independent of whether FLOP counting is active or not. 
- Please add a description of this project to the parent README in `monty_lab`.

File: monty_profiling/frameworks/models/evidence_matching.py (hunk 23 above)
Comment (nielsleadholm) [reply monty_profiling/frameworks/models/evidence_matching.py:107]: That's perfect thanks, yeah I think it would be great to have a comment somewhere (could be in Readme) that for k-d tree, we assume `O(knln_2(n))` for building vs. -- for querying.

File: monty_profiling/frameworks/models/evidence_matching.py (hunk 24 above)
Comment (nielsleadholm) [reply monty_profiling/frameworks/models/evidence_matching.py:123]: Ok interesting thanks - by balanced this presumably means that points are fairly evenly distributed in space? I think that's a reasonable assumption given the structure of our objects, i.e. they don't tend to bunch up in just 2-3 clouds.

File: monty_profiling/frameworks/run_parallel.py [hunk 29]
```diff
@@ -0,0 +1,647 @@
...
+    FlopCountingEvidenceGoalStateGenerator,
+)
+from floppy.monty_flop_tracer import MontyFlopTracer
+
+"""
+Just like run.py, but run episodes in parallel. Running in parallel is as simple as
+
+`python run_parallel.py -e my_exp -n ${NUM_CPUS}`
+
+Assumptions and notes:
```
[the remark this answers is not in the document]
Comment (nielsleadholm) [reply monty_profiling/frameworks/run_parallel.py:50]: Oh wow, yeah that sounds like a (not) fun thing to debug. 

Ok no worries, yeah I think that's absolutely fine, but yeah just worth deleting the `run_parallel.py` code in that case, and adding to the Readme
- A comment that we don't support run_parallel
- A comment to be wary of running in debugging mode (due to slow-downs and infinite recursions)

File: monty_profiling/floppy/monty_flop_tracer.py [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+
+        # Wrap Monty methods
+        wrap_method("_matching_step", self.monty)
+        wrap_method("_exploratory_step", self.monty)
+
+        # Wrap MontyExperiment methods
+        experiment_methods = [
+            "pre_episode",
+            "pre_step",
+            "step",
```
Comment (tristanls) [inline monty_profiling/floppy/monty_flop_tracer.py:91]: **issue**: I think there's a mistake here. This code assumes that `self.experiment.step` method is being wrapped, but the content of `self._original_methods["step"]` is `self.monty.step`. `MontyExperiment` has no `step` method.

Since `run_episode` is wrapped specially, I'm a little lost as to the consequence of this mistake, if any.
```python
    def run_episode(self):
        """Run one episode until model.is_done."""
        self.pre_episode()
        for step, observation in enumerate(self.dataloader):
            self.pre_step(step, observation)
            self.model.step(observation)
            self.post_step(step, observation)
            if self.model.is_done or step >= self.max_steps:
                break
        self.post_episode(step)
```
My question would be, are `self.model.step(observation)` flops included in `self.total_flops`? I don't think `self.model.step` is being wrapped.

Seeing this mistake, it might be better to organize the code into `self._original_monty_methods` and `self._original_experiment_methods`, which would then facilitate something like:

```python
for name in self._original_monty_methods:
  wrap_method(name, self.monty)

for name in self._original_experiment_methods:
  wrap_method(name, self.experiment)
```

**Review by tristanls — COMMENTED** [review]
@nielsleadholm, I was unable to finish reviewing all the changes you highlighted regarding the context manager and the machinery for instrumenting FLOP counting.

This review covers only the following code:
- `TrackedArray`
- `FlopCounter`
- `MontyFlopTracer`
- `monty_profiling/frameworks/run.py`

I restricted my review and comments to double-checking the desired functionality only.

I won't get to the FLOP versions of learning module, goal state, and object model until next week.

**Comment by tristanls** [comment]
I think this pull request would benefit from being redone by creating a new branch from `main` and copying and pasting the new code. Aside from the README update, all the changes appear to be new files. Whatever happened with branch history makes it difficult to review this PR.

File: monty_profiling/floppy/flop_counting/operations/matmul.py (hunk 5 above)
Comment (nielsleadholm) [reply monty_profiling/floppy/flop_counting/operations/matmul.py:85]: Ok, from my perspective, when FLOP counting is active, then that is the primary purpose of the code, and any issue with the program that will cause an incorrect result (like not counting matrix FLOPs) probably should terminate the program. Warnings tend to go unnoticed unless someone is specifically examining the output for issues.

But I'm not sure I understood point (2), is what you're saying that returning None in this case i) might not be that much of an issue, or ii) is unlikely to happen in the first place due to checks elsewhere?

**Comment by nielsleadholm** [comment]
> I think this pull request would benefit from being redone by creating a new branch from `main` and copying and pasting the new code. Aside from the README update, all the changes appear to be new files. Whatever happened with branch history makes it difficult to review this PR.

Thanks for taking an initial look @tristanls . I would be fine with doing this, as long as we keep a copy of this PR active until we've worked through the existing comments.

File: monty_profiling/tests/test_sum.py [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+        _ = a.sum()
+        assert counter.flops == 3
+
+
+def test_sum_axis():
+    counter = FlopCounter()
+    with counter:
+        a = np.array([[1, 2, 3], [4, 5, 6]])
+        _ = np.sum(a, axis=0)  # Sum columns
+        assert counter.flops == 2  # 1 addition per column * 3 columns
```
[the remark this answers is not in the document]
Comment (nielsleadholm) [reply monty_profiling/tests/test_sum.py:28]: Ah ok nice. Yeah I think it's good to have an overview of ones that aren't working, but I wouldn't worry about covering every edge case (e.g. I think it's a reasonable estimate to upper bound sum and ignore axes).

File: monty_profiling/floppy/monty_flop_tracer.py [hunk 32]
```diff
@@ -0,0 +1,115 @@
...
+        self._original_methods = {
+            # Monty methods
+            "_matching_step": self.monty._matching_step,
+            "_exploratory_step": self.monty._exploratory_step,
+            "post_episode": self.monty.post_episode,
+            # MontyExperiment methods
+            "run_episode": self.experiment.run_episode,
+            "pre_episode": self.experiment.pre_episode,
+            "pre_step": self.experiment.pre_step,
+            "step": self.monty.step,
```
[the remark this answers is not in the document]
Comment (tristanls) [reply monty_profiling/floppy/monty_flop_tracer.py]: I was refering to the "step" method.

[the remark this answers is not in the document]
Comment (tristanls) [reply monty_profiling/floppy/monty_flop_tracer.py]: Screenshot 2025-01-21 at 09 02 57
